File: python/n205_SameStruct.py
# https://leetcode-cn.com/problems/isomorphic-strings/

class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        str_map = {}
        for i in range(len(s)):
            str_map[s[i]] = t[i]
        iter = str_map.values()
        if len(list(iter)) != len(set(iter)):
            return False
        p = ""
        for i in range(len(s)):
            p += str_map[s[i]]
        return p == t


if __name__ == "__main__":
    solu = Solution()
    s = "egg"
    t = "add"
    print(solu.isIsomorphic(s, t))

File: cpp/n205_SameStruct.cpp
// https://leetcode-cn.com/problems/isomorphic-strings/
#include <iostream>
#include <vector>
#include <algorithm>
#include <map>
using namespace std;

class Solution {
public:
    string restruct(string s){
        string res = "";
        char symbol = 'a';
        map <char, char> str_map;
        for (size_t i = 0; i < s.size(); ++i){
            if (str_map.find(s[i]) == str_map.end()){
                str_map[s[i]] = symbol;
                symbol += 1;
            }
            res += str_map[s[i]];
        }
        return res;
    }
    
    bool isIsomorphic(string s, string t) {
        return restruct(s) == restruct(t);
    }
};

int main(){
    Solution solu;
    string s = "edg";
    string t = "add";
    cout << solu.isIsomorphic(s, t) << endl;
}